### PR TITLE
feat: add terraform support

### PR DESCRIPTION
**Pull request by lander2k2**

### CODE REVIEW

File: pkg/api/v0/terraform.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	// some terraform resources.
+	TerraformVarsDocument *string `json:"TerraformVarsDocument,omitempty" validate:"optional" encrypt:"true"`
+
+	// The terraform state json object that stores the inventory of
+	// infrastructure being managed by terraform.
+	TerraformStateDocument *string `json:"TerraformStateDocument,omitempty" validate:"optional" encrypt:"true"`
+
+	// The outputs defined in the terraform config that are collected from
+	// Terraform.
+	TerraformOutputs *string `json:"TerraformOutputs,omitempty" validate:"optional" encrypt:"true"`
```
Comment (lander2k2) [inline pkg/api/v0/terraform.go:42]: Add comments to explain why string datatype instead of datatypes.JSON

File: pkg/api/v0/terraform.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	Common         `swaggerignore:"true" mapstructure:",squash"`
+	Instance       `mapstructure:",squash"`
+	Reconciliation `mapstructure:",squash"`
+
+	// The AWS account in which the resources will be provisioned.
+	AwsAccountID *uint `json:"AwsAccountID,omitempty" query:"awsaccountid" gorm:"not null" validate:"required"`
+
+	// The .tfvars document that contains runtime parameters for an instance of
+	// some terraform resources.
+	TerraformVarsDocument *string `json:"TerraformVarsDocument,omitempty" validate:"optional" encrypt:"true"`
```
Comment (lander2k2) [inline pkg/api/v0/terraform.go:34]: Remove Terraform prefix

File: pkg/api/v0/terraform.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// +threeport-codegen:reconciler
+// TerraformDefinition is the configuration for terraform-defined resources.
+type TerraformDefinition struct {
+	Common         `swaggerignore:"true" mapstructure:",squash"`
+	Definition     `mapstructure:",squash"`
+	Reconciliation `mapstructure:",squash"`
+
+	// Path to the directory containing terraform configs with '.tf' file
+	// extension.
+	TerraformConfigDir *string `json:"TerraformConfigDir,omitempty" gorm:"not null" validate:"required"`
```
Comment (lander2k2) [inline pkg/api/v0/terraform.go:14]: Remove Terraform prefix

File: pkg/config/v0/terraform.go [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+// TerraformInstanceConfig contains the config for a terraform instance.
+type TerraformInstanceConfig struct {
+	TerraformInstance TerraformInstanceValues `yaml:"TerraformInstance"`
+}
+
+// TerraformInstanceValues contains the attributes needed to manage a terraform
+// instance.
+type TerraformInstanceValues struct {
+	Name string `yaml:"Name"`
+	//AwsAccountName        string                    `yaml:"AwsAccountName"`
```
Comment (lander2k2) [inline pkg/config/v0/terraform.go:54]: remove

File: internal/terraform/terraform_instance_config.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		fmt.Sprintf("-chdir=%s", c.tfDirName),
+		"apply",
+		"-auto-approve",
+		"-no-color",
+	)
+	applyCmd.Env = append(
+		applyCmd.Environ(),
+		fmt.Sprintf("AWS_ACCESS_KEY_ID=%s", c.accessKeyId),
+		fmt.Sprintf("AWS_SECRET_ACCESS_KEY=%s", c.secretAccessKey),
+	)
```
Comment (lander2k2) [inline internal/terraform/terraform_instance_config.go:64]: See if the terraform CLI supports an AWS_SESSION_TOKEN so that we can use short-lived creds.

**Review by HarisChoudhary — APPROVED** [review]
🚢

**Pull request merged**